File: server/routers/did.py
"""DID API endpoints."""
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from server.dependencies import get_db
from server.schemas.did import (
    DIDCreateRequest, DIDCreateResponse,
    DIDResolveResponse, DIDVerifyRequest, DIDVerifyResponse
)
from server.services.did_service import did_service, DIDNotFoundError, InvalidSignatureError

router = APIRouter(prefix="/did", tags=["DID"])


@router.post("/create", response_model=DIDCreateResponse, status_code=201)
async def create_did(
    request: DIDCreateRequest,
    db: AsyncSession = Depends(get_db),
):
    """Create a new DID with Ed25519 keypair and register the agent."""
    result = await did_service.create(
        db_session=db,
        display_name=request.display_name,
        max_transaction_limit=request.max_transaction_limit,
        allowed_categories=request.allowed_categories,
    )
    return result


@router.get("/resolve/{did:path}", response_model=DIDResolveResponse)
async def resolve_did(did: str):
    """Resolve a DID to its DID Document."""
    try:
        result = await did_service.resolve(did)
        return result
    except DIDNotFoundError:
        raise HTTPException(
            status_code=404,
            detail={"error": {"code": "DID_NOT_FOUND", "message": f"DID not found: {did}", "details": {}}}
        )


@router.post("/verify", response_model=DIDVerifyResponse)
async def verify_did(request: DIDVerifyRequest):
    """Verify an Ed25519 signature against a DID's public key."""
    try:
        await did_service.verify(request.did, request.message, request.signature)
        return {"did": request.did, "verified": True}
    except DIDNotFoundError:
        raise HTTPException(
            status_code=404,
            detail={"error": {"code": "DID_NOT_FOUND", "message": f"DID not found: {request.did}", "details": {}}}
        )
    except InvalidSignatureError:
        raise HTTPException(
            status_code=401,
            detail={"error": {"code": "INVALID_SIGNATURE", "message": "The provided signature does not match the DID's public key.", "details": {}}}
        )

File: server-rust/src/routes/did.rs
use axum::{
    extract::{Path, State},
    http::StatusCode,
    response::IntoResponse,
    routing::{get, post},
    Json, Router,
};
use serde::{Deserialize, Serialize};

use crate::error::AppResult;
use crate::services::did_service;
use crate::state::SharedState;

pub fn router() -> Router<SharedState> {
    Router::new()
        .route("/did/create", post(create_did))
        .route("/did/resolve/*did", get(resolve_did))
        .route("/did/verify", post(verify_did))
}

#[derive(Deserialize)]
struct CreateDIDRequest {
    display_name: Option<String>,
    #[serde(default = "default_max_limit")]
    max_transaction_limit: i64,
    #[serde(default)]
    allowed_categories: Vec<String>,
}

fn default_max_limit() -> i64 {
    100000
}

async fn create_did(
    State(state): State<SharedState>,
    Json(req): Json<CreateDIDRequest>,
) -> AppResult<impl IntoResponse> {
    let result = did_service::create_did(
        &state.db,
        req.display_name,
        req.max_transaction_limit,
        req.allowed_categories,
    )
    .await?;
    Ok((StatusCode::CREATED, Json(result)))
}

async fn resolve_did(
    State(state): State<SharedState>,
    Path(did): Path<String>,
) -> AppResult<impl IntoResponse> {
    let result = did_service::resolve_did(&state.db, &did).await?;
    Ok(Json(result))
}

#[derive(Deserialize)]
struct VerifyDIDRequest {
    did: String,
    message: String,
    signature: String,
}

async fn verify_did(
    State(state): State<SharedState>,
    Json(req): Json<VerifyDIDRequest>,
) -> AppResult<impl IntoResponse> {
    did_service::verify_did(&state.db, &req.did, &req.message, &req.signature).await?;
    Ok(Json(serde_json::json!({
        "did": req.did,
        "verified": true,
    })))
}
